osu.Game/Extensions/DrawableExtensions.cs

```diff
@@ -79,5 +80,42 @@ public static void ApplySkinnableInfo(this Drawable component, SkinnableInfo inf
                     container.Add(child.CreateInstance());
             }
         }
+
+        /// <summary>
+        /// Keeps the drawable upright and prevents it from being scaled or flipped with its Parent.
+        /// </summary>
+        /// <param name="drawable">The drawable.</param>
+        public static void KeepUprightAndUnscaled(this Drawable drawable)
+        {
+            var parentMatrix = drawable.Parent.DrawInfo.Matrix;
+            float angle = MathF.Atan(parentMatrix.M12 / parentMatrix.M11);
+            angle = MathHelper.RadiansToDegrees(angle);
+
+            parentMatrix.Transpose();
+            parentMatrix.M13 = 0.0f;
+            parentMatrix.M23 = 0.0f;
+
+            if ((Math.Abs(Math.Abs(angle) - 90.0)) < 2.0f)
+            {
+                Matrix3 m = Matrix3.CreateRotationZ(MathHelper.DegreesToRadians(40.0f));
+                m.Transpose();
+                parentMatrix *= m;
+                drawable.Rotation = 40.0f;
+            }
```

Comment:
What's happening here? What's the check for, why is the threshold 2? Why 40 degrees?

Reply:
As i mentioned in my reply below this implementation applies shear and scale to the drawable to revert the rotation and scale applied to the parent container, but it we can only revert the rotation and scale applied to the parent container with a shear and scale if the rotation is not equal to (90 or -90), in order to work around that we rotate the drawable a certain amount (it does not matter how much we rotate it all that matters is that we do not rotate it a multiple of 180 and i picked the number 40 randomly), and we adjust parentMatrix we that rotation.

In line 98 we are not only checking if the angle applied to the parent is not 90 or -90, we are also avoiding another problem, that is the more the rotation approaches 90 or -90 the bigger or smaller the scale and shear factors get, which causes a bug (the drawable disappears, and its update function does not get called again), i think that the drawable disappears because the factors being too big and too small is causing the floating point operations to be inaccurate.

I introduced a threshold to avoid that bug, i could have picked 1 degrees (or any small number) as the threshold but i picked 2 degrees just to be safe.